File: app/src/main/java/com/example/domain/ai/model/AiCapability.kt
package com.example.domain.ai.model

/**
 * As capabilities de IA que o backend autoriza por conta (T19.0).
 *
 * Espelha `AI_CAPABILITIES` em `backend/src/modules/ai/entitlement/ai-capability.ts` — os dois
 * lados são o mesmo contrato, do mesmo jeito que [AiCoachRequestType] já espelha
 * `AI_COACH_REQUEST_TYPES`. O Android nunca decide entitlement: ele só lê o que o servidor
 * respondeu, para construir uma UX coerente. Toda chamada real continua sendo autorizada de novo
 * no backend.
 */
enum class AiCapability {
    AI_ANALYZE_WORKOUT,
    AI_GENERATE_WORKOUT,
    AI_ADAPT_WORKOUT,
    AI_EXPLAIN
}

/**
 * A capability exigida por esta operação do Coach (T19.0 §6.5) — o mesmo mapeamento de
 * `capabilityFor()` no backend, mantido separado do backend de propósito: os dois lados precisam
 * concordar, mas nenhum importa o outro.
 *
 * Um [AiCoachRequestType] novo sem entrada aqui não compila: `when` sobre um enum é exaustivo no
 * Kotlin, e é isso que impede uma operação nova cair silenciosamente numa capability genérica.
 */
val AiCoachRequestType.requiredCapability: AiCapability
    get() = when (this) {
        AiCoachRequestType.ANALYZE_WORKOUT -> AiCapability.AI_ANALYZE_WORKOUT
        AiCoachRequestType.GENERATE_WORKOUT -> AiCapability.AI_GENERATE_WORKOUT
        AiCoachRequestType.ADAPT_WORKOUT -> AiCapability.AI_ADAPT_WORKOUT
        AiCoachRequestType.EXPLAIN_RECOMMENDATION,
        AiCoachRequestType.EXPLAIN_WORKOUT,
        AiCoachRequestType.EXPLAIN_ADAPTATION,
        AiCoachRequestType.EXPLAIN_PROGRESS -> AiCapability.AI_EXPLAIN
    }

/** Por que uma consulta de capabilities não produziu um resultado utilizável. */
enum class AiCapabilitiesErrorKind {
    /** Sem conta conectada — capabilities de IA exigem Conta Spark, como o resto do Coach. */
    AUTH_REQUIRED,
    /** Backend indisponível, não configurado neste build, ou o servidor não conseguiu decidir. */
    UNAVAILABLE,
    /** Sem conectividade ou falha de transporte. */
    NETWORK,
    /** A resposta chegou, mas não pôde ser interpretada com segurança. */
    INVALID_RESPONSE
}

/** O resultado bruto de `GET /v1/account/capabilities`. */
sealed interface AiCapabilitiesGatewayResult {
    /** As capabilities que o servidor confirmou liberadas. Ausência = não liberada. */
    data class Success(val allowed: Set<AiCapability>) : AiCapabilitiesGatewayResult
    data class Error(val kind: AiCapabilitiesErrorKind) : AiCapabilitiesGatewayResult
}

File: backend/src/modules/ai/entitlement/ai-capability.ts
import { isExplanation, type AiCoachRequestType } from '../ai-coach.contract';

/**
 * As capabilities de IA que a T19.0 introduz.
 *
 * Cada uma responde a uma pergunta binária — "esta conta pode usar esta funcionalidade?" — e
 * nenhuma delas é um plano. Uma capability nova precisa entrar aqui **e** em
 * `AiCapability.kt` no Android; os dois lados são o mesmo contrato, do mesmo jeito que
 * `AI_COACH_REQUEST_TYPES` já é espelhado nos dois lados.
 */
export const AI_CAPABILITIES = [
  'AI_ANALYZE_WORKOUT',
  'AI_GENERATE_WORKOUT',
  'AI_ADAPT_WORKOUT',
  'AI_EXPLAIN',
] as const;

export type AiCapability = (typeof AI_CAPABILITIES)[number];

export function isAiCapability(value: string): value is AiCapability {
  return (AI_CAPABILITIES as readonly string[]).includes(value);
}

/**
 * Operação do Coach → capability exigida.
 *
 * Único lugar com este mapeamento (§6.5): os quatro `EXPLAIN_*` reaproveitam o discriminador
 * `isExplanation()` que o contrato já tem, em vez de repetir a lista de quatro nomes aqui. Uma
 * capability nova para uma nova operação entra numa correspondência 1:1 explícita — este `switch`
 * nunca tem um `default` que devolve uma capability genérica, porque isso é exatamente o "cair
 * silenciosamente em permissão genérica" que a tarefa proíbe (§11).
 */
export function capabilityFor(requestType: AiCoachRequestType): AiCapability {
  if (isExplanation(requestType)) {
    return 'AI_EXPLAIN';
  }
  switch (requestType) {
    case 'ANALYZE_WORKOUT':
      return 'AI_ANALYZE_WORKOUT';
    case 'GENERATE_WORKOUT':
      return 'AI_GENERATE_WORKOUT';
    case 'ADAPT_WORKOUT':
      return 'AI_ADAPT_WORKOUT';
    default: {
      // Exaustividade em tempo de compilação: um `AiCoachRequestType` novo sem entrada aqui quebra
      // o build, em vez de silenciosamente herdar uma capability errada.
      const exhaustive: never = requestType;
      throw new Error(`operação de Coach sem capability mapeada: ${String(exhaustive)}`);
    }
  }
}
